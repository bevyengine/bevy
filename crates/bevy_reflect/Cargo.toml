[package]
name = "bevy_reflect"
version = "0.4.0"
edition = "2018"
authors = [
    "Bevy Contributors <bevyengine@gmail.com>",
    "Carter Anderson <mcanders1@gmail.com>",
]
description = "Dynamically interact with rust types"
homepage = "https://bevyengine.org"
repository = "https://github.com/bevyengine/bevy"
license = "MIT"
keywords = ["bevy"]
readme = "README.md"

[features]
bevy = ["bevy_ecs", "bevy_app", "glam", "smallvec"]

[dependencies]
# bevy
bevy_reflect_derive = { path = "bevy_reflect_derive", version = "0.4.0" }
bevy_utils = { path = "../bevy_utils", version = "0.4.0" }
bevy_ecs = { path = "../bevy_ecs", version = "0.4.0", optional = true }
bevy_app = { path = "../bevy_app", version = "0.4.0", optional = true }

# other
erased-serde = "0.3"
downcast-rs = "1.2"
parking_lot = "0.11.0"
thiserror = "1.0"
serde = "1"
smallvec = { version = "1.4", features = ["serde"], optional = true }
glam = { version = "0.13.0", features = ["serde"], optional = true }

[dev-dependencies]
ron = "0.6.2"
