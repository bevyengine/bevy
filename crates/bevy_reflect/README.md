# Bevy Reflect

This crate enables you to dynamically interact with Rust types:

* Derive the Reflect traits
* Interact with fields using their names (for named structs) or indices (for tuple structs)
* "Patch" your types with new values
* Look up nested fields using "path strings"
* Iterate over struct fields
* Automatically serialize and deserialize via Serde (without explicit serde impls)
* Trait "reflection" 

## Features

### Derive the Reflect traits

```rust
// this will automatically implement the Reflect trait and the Struct trait (because the type is a struct)
#[derive(Reflect)]
struct Foo {
    a: u32,
    b: Bar
    c: Vec<i32>,
    d: Vec<Bar>,
}

// this will automatically implement the Reflect trait and the TupleStruct trait (because the type is a tuple struct)
#[derive(Reflect)]
struct Bar(String);

#[derive(Reflect)]
struct Baz {
    value: f32,
};

// We will use this value to illustrate `bevy_reflect` features
let mut foo = Foo {
    a: 1,
    b: Bar("hello".to_string()),
    c: vec![1, 2]
    d: vec![Baz { value: 3.14 }]
};
```

### Interact with fields using their names

```rust
assert_eq!(*foo.get_field::<u32>("a").unwrap(), 1);

*foo.get_field_mut::<u32>("a").unwrap() = 2;

assert_eq!(foo.a, 2);
```

### "Patch" your types with new values

```rust
let mut dynamic_struct = DynamicStruct::default();
dynamic_struct.insert("a", 42u32);
dynamic_struct.insert("c", vec![3, 4, 5]);

foo.apply(&dynamic_struct);

assert_eq!(foo.a, 42);
assert_eq!(foo.c, vec![3, 4, 5]);
```

### Look up nested fields using "path strings"

```rust
let value = *foo.get_path::<f32>("d[0].value").unwrap();
assert_eq!(value, 3.14);
```

### Iterate over struct fields

```rust
for (i, value: &Reflect) in foo.iter_fields().enumerate() {
    let field_name = foo.name_at(i).unwrap();
    if let Ok(value) = value.downcast_ref::<u32>() {
        println!("{} is a u32 with the value: {}", field_name, *value);
    } 
}
```

### Automatically serialize and deserialize via Serde (without explicit serde impls)

```rust
let mut registry = TypeRegistry::default();
registry.register::<u32>();
registry.register::<i32>();
registry.register::<f32>();
registry.register::<String>();
registry.register::<Bar>();
registry.register::<Baz>();

let serializer = ReflectSerializer::new(&foo, &registry);
let serialized = ron::ser::to_string_pretty(&serializer, ron::ser::PrettyConfig::default()).unwrap();

let mut deserializer = ron::de::Deserializer::from_str(&serialized).unwrap();
let reflect_deserializer = ReflectDeserializer::new(&registry);
let value = reflect_deserializer.deserialize(&mut deserializer).unwrap();
let dynamic_struct = value.take::<DynamicStruct>().unwrap();

assert!(foo.partial_eq(&dynamic_struct).unwrap());
```

### Trait "reflection"

Call a trait on a given &dyn Reflect reference without knowing the underlying type!

```rust
#[derive(Reflect)]
#[reflect(DoThing)]
struct MyType {
    value: String,
}

impl DoThing for MyType {
    fn do_thing(&self) -> String {
        format!("{} World!", self.value)
    }
}

#[reflect_trait]
pub trait DoThing {
    fn do_thing(&self) -> String;
}

// First, lets box our type as a Box<dyn Reflect>
let reflect_value: Box<dyn Reflect> = Box::new(MyType {
    value: "Hello".to_string(),
});

// This means we no longer have direct access to MyType or its methods. We can only call Reflect methods on reflect_value.
// What if we want to call `do_thing` on our type? We could downcast using reflect_value.downcast_ref::<MyType>(), but what if we
// don't know the type at compile time?

// Normally in rust we would be out of luck at this point. Lets use our new reflection powers to do something cool!
let mut type_registry = TypeRegistry::default()
type_registry.register::<MyType>();

// The #[reflect] attribute we put on our DoThing trait generated a new `ReflectDoThing` struct, which implements TypeData.
// This was added to MyType's TypeRegistration.
let reflect_do_thing = type_registry
    .get_type_data::<ReflectDoThing>(reflect_value.type_id())
    .unwrap();

// We can use this generated type to convert our `&dyn Reflect` reference to a `&dyn DoThing` reference
let my_trait: &dyn DoThing = reflect_do_thing.get(&*reflect_value).unwrap();

// Which means we can now call do_thing(). Magic!
println!("{}", my_trait.do_thing());

// This works because the #[reflect(MyTrait)] we put on MyType informed the Reflect derive to insert a new instance
// of ReflectDoThing into MyType's registration. The instance knows how to cast &dyn Reflect to &dyn MyType, because it
// knows that &dyn Reflect should first be downcasted to &MyType, which can then be safely casted to &dyn MyType
```

## Why make this?

The whole point of Rust is static safety! Why build something that makes it easy to throw it all away?

* Some problems are inherently dynamic (scripting, some types of serialization / deserialization)
* Sometimes the dynamic way is easier
* Sometimes the dynamic way puts less burden on your users to derive a bunch of traits (this was a big motivator for the Bevy project)