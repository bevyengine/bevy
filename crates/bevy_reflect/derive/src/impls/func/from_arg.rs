use crate::{derive_data::ReflectMeta, where_clause_options::WhereClauseOptions};
use bevy_macro_utils::fq_std::FQResult;
use quote::quote;

pub(crate) fn impl_from_arg(
    meta: &ReflectMeta,
    where_clause_options: &WhereClauseOptions,
) -> proc_macro2::TokenStream {
    let bevy_reflect = meta.bevy_reflect_path();
    let type_path = meta.type_path();

    let (impl_generics, ty_generics, where_clause) = type_path.generics().split_for_impl();
    let where_reflect_clause = where_clause_options.extend_where_clause(where_clause);

    quote! {
        impl #impl_generics #bevy_reflect::func::args::FromArg for #type_path #ty_generics #where_reflect_clause {
            type This<'from_arg> = #type_path #ty_generics;
            fn from_arg(arg: #bevy_reflect::func::args::Arg) -> #FQResult<Self::This<'_>, #bevy_reflect::func::args::ArgError> {
                arg.take_owned()
            }
        }

        impl #impl_generics #bevy_reflect::func::args::FromArg for &'static #type_path #ty_generics #where_reflect_clause {
            type This<'from_arg> = &'from_arg #type_path #ty_generics;
            fn from_arg(arg: #bevy_reflect::func::args::Arg) -> #FQResult<Self::This<'_>, #bevy_reflect::func::args::ArgError> {
                arg.take_ref()
            }
        }

        impl #impl_generics #bevy_reflect::func::args::FromArg for &'static mut #type_path #ty_generics #where_reflect_clause {
            type This<'from_arg> = &'from_arg mut #type_path #ty_generics;
            fn from_arg(arg: #bevy_reflect::func::args::Arg) -> #FQResult<Self::This<'_>, #bevy_reflect::func::args::ArgError> {
                arg.take_mut()
            }
        }
    }
}
