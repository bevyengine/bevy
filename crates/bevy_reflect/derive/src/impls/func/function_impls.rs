use crate::{
    derive_data::ReflectMeta,
    impls::func::{
        from_arg::impl_from_arg, get_ownership::impl_get_ownership, into_return::impl_into_return,
    },
    where_clause_options::WhereClauseOptions,
};
use quote::quote;

pub(crate) fn impl_function_traits(
    meta: &ReflectMeta,
    where_clause_options: &WhereClauseOptions,
) -> proc_macro2::TokenStream {
    let get_ownership = impl_get_ownership(meta, where_clause_options);
    let from_arg = impl_from_arg(meta, where_clause_options);
    let into_return = impl_into_return(meta, where_clause_options);

    quote! {
        #get_ownership

        #from_arg

        #into_return
    }
}
