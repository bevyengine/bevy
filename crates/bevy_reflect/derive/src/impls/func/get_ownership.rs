use crate::{derive_data::ReflectMeta, where_clause_options::WhereClauseOptions};
use quote::quote;

pub(crate) fn impl_get_ownership(
    meta: &ReflectMeta,
    where_clause_options: &WhereClauseOptions,
) -> proc_macro2::TokenStream {
    let bevy_reflect = meta.bevy_reflect_path();
    let type_path = meta.type_path();

    let (impl_generics, ty_generics, where_clause) = type_path.generics().split_for_impl();
    let where_reflect_clause = where_clause_options.extend_where_clause(where_clause);

    quote! {
        impl #impl_generics #bevy_reflect::func::args::GetOwnership for #type_path #ty_generics #where_reflect_clause {
            fn ownership() -> #bevy_reflect::func::args::Ownership {
                #bevy_reflect::func::args::Ownership::Owned
            }
        }

        impl #impl_generics #bevy_reflect::func::args::GetOwnership for &'_ #type_path #ty_generics #where_reflect_clause {
            fn ownership() -> #bevy_reflect::func::args::Ownership {
                #bevy_reflect::func::args::Ownership::Ref
            }
        }

        impl #impl_generics #bevy_reflect::func::args::GetOwnership for &'_ mut #type_path #ty_generics #where_reflect_clause {
            fn ownership() -> #bevy_reflect::func::args::Ownership {
                #bevy_reflect::func::args::Ownership::Mut
            }
        }
    }
}
