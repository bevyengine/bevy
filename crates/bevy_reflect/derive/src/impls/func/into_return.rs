use crate::derive_data::ReflectMeta;
use crate::utility::WhereClauseOptions;
use quote::quote;

pub(crate) fn impl_into_return(
    meta: &ReflectMeta,
    where_clause_options: &WhereClauseOptions,
) -> proc_macro2::TokenStream {
    let bevy_reflect = meta.bevy_reflect_path();
    let type_path = meta.type_path();

    let (impl_generics, ty_generics, where_clause) = type_path.generics().split_for_impl();
    let where_reflect_clause = where_clause_options.extend_where_clause(where_clause);

    quote! {
        impl #impl_generics #bevy_reflect::func::IntoReturn for #type_path #ty_generics #where_reflect_clause {
            fn into_return<'into_return>(self) -> #bevy_reflect::func::Return<'into_return> {
                #bevy_reflect::func::Return::Owned(Box::new(self))
            }
        }

        impl #impl_generics #bevy_reflect::func::IntoReturn for &'static #type_path #ty_generics #where_reflect_clause {
            fn into_return<'into_return>(self) -> #bevy_reflect::func::Return<'into_return> {
                #bevy_reflect::func::Return::Ref(self)
            }
        }

        impl #impl_generics #bevy_reflect::func::IntoReturn for &'static mut #type_path #ty_generics #where_reflect_clause {
            fn into_return<'into_return>(self) -> #bevy_reflect::func::Return<'into_return> {
                #bevy_reflect::func::Return::Mut(self)
            }
        }
    }
}
