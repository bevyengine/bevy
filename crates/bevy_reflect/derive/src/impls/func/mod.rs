pub(crate) use function_impls::impl_function_traits;

mod from_arg;
mod function_impls;
mod get_ownership;
mod into_return;
