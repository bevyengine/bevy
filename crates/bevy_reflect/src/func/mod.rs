//! Reflection-based dynamic functions.
//!
//! This module provides a way to pass around and call functions dynamically
//! using the [`DynamicFunction`] type.
//!
//! Many simple functions and closures can be automatically converted to [`DynamicFunction`]
//! using the [`IntoFunction`] trait.
//!
//! Once the [`DynamicFunction`] is created, it can be called with a set of arguments provided
//! via an [`ArgList`].
//!
//! This returns a [`FunctionResult`] containing the [`Return`] value,
//! which can be used to extract a [`Reflect`] trait object.
//!
//!
//! # Example
//!
//! ```
//! # use bevy_reflect::Reflect;
//! # use bevy_reflect::func::args::ArgList;
//! # use bevy_reflect::func::{DynamicFunction, FunctionResult, IntoFunction, Return};
//! fn add(a: i32, b: i32) -> i32 {
//!   a + b
//! }
//!
//! let mut func: DynamicFunction = add.into_function();
//! let args: ArgList = ArgList::default()
//!   // Pushing a known type with owned ownership
//!   .push_owned(25_i32)
//!   // Pushing a reflected type with owned ownership
//!   .push_boxed(Box::new(75_i32) as Box<dyn Reflect>);
//! let result: FunctionResult = func.call(args);
//! let value: Return = result.unwrap();
//! assert_eq!(value.unwrap_owned().downcast_ref::<i32>(), Some(&100));
//! ```
//!
//! [`Reflect`]: crate::Reflect

pub use error::*;
pub use function::*;
pub use info::*;
pub use into_function::*;
pub use return_type::*;

pub use args::{Arg, ArgError, ArgList};

pub mod args;
mod error;
mod function;
mod info;
mod into_function;
pub(crate) mod macros;
mod return_type;

#[cfg(test)]
mod tests {
    use super::*;
    use crate as bevy_reflect;
    use crate::func::args::{ArgError, ArgId, ArgList, Ownership};
    use crate::{Reflect, TypePath};
    use alloc::borrow::Cow;

    #[test]
    fn should_create_dynamic_function() {
        fn add(a: i32, b: i32) -> i32 {
            a + b
        }

        let mut func = add.into_function();
        let args = ArgList::new().push_owned(25_i32).push_owned(75_i32);
        let result = func.call(args).unwrap().unwrap_owned();
        assert_eq!(result.downcast_ref::<i32>(), Some(&100));
    }

    #[test]
    fn should_create_dynamic_closure() {
        let mut func = (|a: i32, b: i32| a + b).into_function();
        let args = ArgList::new().push_owned(25_i32).push_owned(75_i32);
        let result = func.call(args).unwrap().unwrap_owned();
        assert_eq!(result.downcast_ref::<i32>(), Some(&100));
    }

    #[test]
    fn should_create_dynamic_method() {
        #[derive(Reflect, Debug, PartialEq)]
        struct Foo(i32);

        impl Foo {
            pub fn add(&self, other: &Foo) -> Foo {
                Foo(self.0 + other.0)
            }
        }

        let foo_a = Foo(25);
        let foo_b = Foo(75);

        let mut func = Foo::add.into_function();
        let args = ArgList::new().push_ref(&foo_a).push_ref(&foo_b);
        let result = func.call(args).unwrap().unwrap_owned();
        assert_eq!(result.downcast_ref::<Foo>(), Some(&Foo(100)));
    }

    #[test]
    fn should_allow_zero_args() {
        fn foo() -> String {
            String::from("Hello, World!")
        }

        let mut func = foo.into_function();
        let args = ArgList::new();
        let result = func.call(args).unwrap().unwrap_owned();
        assert_eq!(
            result.downcast_ref::<String>(),
            Some(&String::from("Hello, World!"))
        );
    }

    #[test]
    fn should_allow_unit_return() {
        fn foo(_: i32) {}

        let mut func = foo.into_function();
        let args = ArgList::new().push_owned(123_i32);
        let result = func.call(args).unwrap();
        assert!(result.is_unit());
    }

    #[test]
    fn should_allow_reference_return() {
        fn foo<'a>(value: &'a i32, _: String, _: &bool) -> &'a i32 {
            value
        }

        let value: i32 = 123;
        let mut func = foo.into_function();
        let args = ArgList::new()
            .push_ref(&value)
            .push_owned(String::from("Hello, World!"))
            .push_ref(&true);
        let result = func.call(args).unwrap().unwrap_ref();
        assert_eq!(result.downcast_ref::<i32>(), Some(&123));
    }

    #[test]
    fn should_allow_mutable_reference_return() {
        fn foo<'a>(value: &'a mut i32, _: String, _: &bool) -> &'a mut i32 {
            value
        }

        let mut value: i32 = 123;
        let mut func = foo.into_function();
        let args = ArgList::new()
            .push_mut(&mut value)
            .push_owned(String::from("Hello, World!"))
            .push_ref(&true);
        let result = func.call(args).unwrap().unwrap_mut();
        assert_eq!(result.downcast_mut::<i32>(), Some(&mut 123));
    }

    #[test]
    fn should_default_with_function_type_name() {
        fn foo() {}

        let func = foo.into_function();
        assert_eq!(
            func.info().name(),
            Some("bevy_reflect::func::tests::should_default_with_function_type_name::foo")
        );
    }

    #[test]
    fn should_default_with_closure_type_name() {
        let bar = |_: i32| {};

        let func = bar.into_function();
        assert_eq!(
            func.info().name(),
            Some("bevy_reflect::func::tests::should_default_with_closure_type_name::{{closure}}")
        );
    }

    #[test]
    fn should_overwrite_function_name() {
        fn foo() {}

        let func = foo.into_function().with_name("my_function");
        assert_eq!(func.info().name(), Some("my_function"));
    }

    #[test]
    fn should_error_on_missing_args() {
        fn foo(_: i32) {}

        let mut func = foo.into_function();
        let args = ArgList::new();
        let result = func.call(args);
        assert_eq!(
            result.unwrap_err(),
            FunctionError::InvalidArgCount {
                expected: 1,
                received: 0
            }
        );
    }

    #[test]
    fn should_error_on_too_many_args() {
        fn foo() {}

        let mut func = foo.into_function();
        let args = ArgList::new().push_owned(123_i32);
        let result = func.call(args);
        assert_eq!(
            result.unwrap_err(),
            FunctionError::InvalidArgCount {
                expected: 0,
                received: 1
            }
        );
    }

    #[test]
    fn should_error_on_invalid_arg_type() {
        fn foo(_: i32) {}

        let mut func = foo.into_function();
        let args = ArgList::new().push_owned(123_u32);
        let result = func.call(args);
        assert_eq!(
            result.unwrap_err(),
            FunctionError::ArgError(ArgError::UnexpectedType {
                id: ArgId::Index(0),
                expected: Cow::Borrowed(i32::type_path()),
                received: Cow::Borrowed(u32::type_path())
            })
        );
    }

    #[test]
    fn should_error_on_invalid_arg_ownership() {
        fn foo(_: &i32) {}

        let mut func = foo.into_function();
        let args = ArgList::new().push_owned(123_i32);
        let result = func.call(args);
        assert_eq!(
            result.unwrap_err(),
            FunctionError::ArgError(ArgError::InvalidOwnership {
                id: ArgId::Index(0),
                expected: Ownership::Ref,
                received: Ownership::Owned
            })
        );
    }

    #[test]
    fn should_convert_dynamic_function_with_into_function() {
        fn make_function<'a, F: IntoFunction<'a, M>, M>(f: F) -> DynamicFunction<'a> {
            f.into_function()
        }

        let function: DynamicFunction = make_function(|| {});
        let _: DynamicFunction = make_function(function);
    }
}
