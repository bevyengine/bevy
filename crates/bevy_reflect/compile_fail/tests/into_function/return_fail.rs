#![allow(unused)]

use bevy_reflect::func::IntoFunction;
use bevy_reflect::Reflect;

fn pass() -> i32 {
    123
}

struct Foo;

fn return_not_reflect() -> Foo {
    Foo
}

fn return_with_lifetime_pass<'a>(a: &'a String) -> &'a String {
    a
}

fn return_with_invalid_lifetime<'a, 'b>(a: &'a String, b: &'b String) -> &'b String {
    b
}

fn main() {
    let _ = pass.into_function();

    let _ = return_not_reflect.into_function();
    //~^ E0599

    let _ = return_with_lifetime_pass.into_function();

    let _ = return_with_invalid_lifetime.into_function();
    //~^ E0599
}
