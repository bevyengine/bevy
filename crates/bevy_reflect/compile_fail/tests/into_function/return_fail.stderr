error[E0599]: the method `into_function` exists for fn item `fn() -> Foo {return_not_reflect}`, but its trait bounds were not satisfied
  --> tests/into_function/return_fail.rs:LL:CC
   |
LL |     let _ = return_not_reflect.into_function();
   |                                ^^^^^^^^^^^^^ method cannot be called on `fn() -> Foo {return_not_reflect}` due to unsatisfied trait bounds
   |
   = note: the following trait bounds were not satisfied:
           `fn() -> Foo {return_not_reflect}: ReflectFn<'_, _>`
           which is required by `fn() -> Foo {return_not_reflect}: IntoFunction<'_, (_, _)>`
           `fn() -> Foo {return_not_reflect}: TypedFunction<_>`
           which is required by `fn() -> Foo {return_not_reflect}: IntoFunction<'_, (_, _)>`
           `&fn() -> Foo {return_not_reflect}: ReflectFn<'_, _>`
           which is required by `&fn() -> Foo {return_not_reflect}: IntoFunction<'_, (_, _)>`
           `&fn() -> Foo {return_not_reflect}: TypedFunction<_>`
           which is required by `&fn() -> Foo {return_not_reflect}: IntoFunction<'_, (_, _)>`
           `&mut fn() -> Foo {return_not_reflect}: ReflectFn<'_, _>`
           which is required by `&mut fn() -> Foo {return_not_reflect}: IntoFunction<'_, (_, _)>`
           `&mut fn() -> Foo {return_not_reflect}: TypedFunction<_>`
           which is required by `&mut fn() -> Foo {return_not_reflect}: IntoFunction<'_, (_, _)>`

error[E0599]: the method `into_function` exists for fn item `fn(&String, &String) -> &String {return_with_invalid_lifetime}`, but its trait bounds were not satisfied
  --> tests/into_function/return_fail.rs:LL:CC
   |
LL |     let _ = return_with_invalid_lifetime.into_function();
   |                                          ^^^^^^^^^^^^^ method cannot be called due to unsatisfied trait bounds
   |
   = note: the full type name has been written to 'BEVY_ROOT/bevy_reflect/compile_fail/target/ui/0/tests/into_function/long-type.sr''
   = note: consider using `--verbose` to print the full type name to the console
   = note: the following trait bounds were not satisfied:
           `for<'a, 'b> fn(&'a std::string::String, &'b std::string::String) -> &'b std::string::String {return_with_invalid_lifetime}: ReflectFn<'_, _>`
           which is required by `for<'a, 'b> fn(&'a std::string::String, &'b std::string::String) -> &'b std::string::String {return_with_invalid_lifetime}: IntoFunction<'_, (_, _)>`
           `&for<'a, 'b> fn(&'a std::string::String, &'b std::string::String) -> &'b std::string::String {return_with_invalid_lifetime}: ReflectFn<'_, _>`
           which is required by `&for<'a, 'b> fn(&'a std::string::String, &'b std::string::String) -> &'b std::string::String {return_with_invalid_lifetime}: IntoFunction<'_, (_, _)>`
           `&mut for<'a, 'b> fn(&'a std::string::String, &'b std::string::String) -> &'b std::string::String {return_with_invalid_lifetime}: ReflectFn<'_, _>`
           which is required by `&mut for<'a, 'b> fn(&'a std::string::String, &'b std::string::String) -> &'b std::string::String {return_with_invalid_lifetime}: IntoFunction<'_, (_, _)>`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0599`.
